File: src/step1_update_pmi/convert_frequencies_to_pPMI.py
# This is a new version of an old program-- this version uses the latest counts of
# target and context words from the Easton participants.

import pickle
from math import log, exp
import numpy as np
from numpy.linalg import norm

def pluralize(w):
    if w[-1] == 's' or w[-1] == 'z' or w[-2:] == 'sh':
        return w + 'es'
    if w[-1] == 'y':
        return w[:-1] + 'ies'
    return w + 's'

print "Loading joint frequencies..."
fh = open('contextVectors.pkl','r')
joints = pickle.load(fh)
fh.close()

print "Combining vectors for inflected targets..."
def find_inflected(ks):
    tups = []
    for k in ks:
        infl = pluralize(k)
        if infl in ks:
            tups.append((k,infl))
    return tups

inflected = find_inflected([k for k in joints.keys() if k])
print inflected

def combine_inflected(hash,tups):
    for (un,inf) in tups:
        for key in hash[inf]:
            if not hash[un].has_key(key):
                hash[un][key] = hash[inf][key]
            else:
                hash[un][key] += hash[inf][key]
    return hash

joints = combine_inflected(joints,inflected)

print "Loading raw frequencies..."
fh = open('correctedLogUnigrams.pkl','r')
raws = pickle.load(fh)
fh.close()
raws = dict([ (k,exp(raws[k])) for k in raws ])
total_raw = float(sum(raws.values()))
raw_prob = dict([ (key,raws[key]/total_raw) for key in raws ])
raws = None

print "Calculating probability of context words..."
context_prob = {}
for target in joints:
    sum_contexts = float(sum(joints[target].values()))
    context_prob[target] = {}
    for con in joints[target].keys():
        context_prob[target][con] = joints[target][con]/sum_contexts

print "Calculating positive pointwise mutual information for entries in main hash table..."
# Using formula pPMI = max(0,log(P(context|targ)/P(context)))
pPMI = {}
for target in context_prob.keys():
    pPMI[target] = {}
    for con in context_prob[target].keys():
        pConGivenTarg = context_prob[target][con]
        try:
            pCon = raw_prob[con]
        except KeyError:
            pCon = 0
        try:
            ppmi = log(pConGivenTarg) - log(pCon)
        except ValueError:
            ppmi = 0
        pPMI[target][con] = max(0,ppmi)


print "Saving pPMI to file pPMI_all.pickle"
fh = open("pPMI_all.pickle",'w')
pickle.dump(pPMI,fh)
fh.close()

def cosim_hash(h1,h2):
    allKeys = sorted(list(set(h1.keys()).union(set(h2.keys()))))
    v1 = np.zeros(len(allKeys))
    v2 = np.zeros(len(allKeys))
    for (i,k) in enumerate(allKeys):
        if h1.has_key(k):
            v1[i] = h1[k]
        if h2.has_key(k):
            v2[i] = h2[k]
    n1 = norm(v1)
    n2 = norm(v2)
    if n1 != 0 and n2 != 0:
        v1 = v1/n1
        v2 = v2/n2
        return np.dot(v1,v2)
    else:
        return 0.0

def top_items(s):
    coses = [(cosim_hash(pPMI[k],pPMI[s]),k) for k in pPMI.keys() if k != s]
    thetas = sorted([ c for (c,k) in coses ])
    theta = thetas[-15]
    return sorted([ (c,k) for (c,k) in coses if c >= theta ],reverse=True)

for word in [ 'eagle', 'socks', 'jacket', 'asparagus', 'brussels_sprouts', 'spaghetti', 'lion', 'fun', 'arrive', 'straight' ]:
    print word, ':'
    top15 = top_items(word)
    for (co,item) in top15:
        print '\t', item, co

        

File: src/step1_update_pmi/convertFrequenciesToPpmi.jl
using JSON
using Lazy

typealias StringIntMap Dict{ASCIIString, Int64}
typealias JointFrequency Dict{ASCIIString, StringIntMap}

typealias StringFloatMap Dict{ASCIIString, Float64}
typealias JointProbability Dict{ASCIIString, StringFloatMap}


jointFrequencies(context_vectors_path::AbstractString) = JSON.parsefile(
  context_vectors_path, dicttype=JointFrequency)


rawProb(unigramprob_f::AbstractString) = rawProb(
  JSON.parsefile(unigramprob_f, dicttype=StringFloatMap))


jointProbabilities(context_vectors_path::AbstractString) = jointProbabilities(
  jointFrequencies(context_vectors_path))


function jointProbabilities(joint_freq::JointFrequency)
  joint_prob = JointProbability()

  for (target, context_counts) in joint_freq
    sum_contexts = context_counts |> values |> sum
    joint_prob[target] = StringFloatMap()
    for (context, context_count) in context_counts
      joint_prob[target][context] = context_count/sum_contexts
    end
  end

  joint_prob
end


function pPmi(context_vectors_path::AbstractString, logunigrams_f::AbstractString)
  pPmi(jointProbabilities(context_vectors_path), rawProb(logunigrams_f))
end


function pPmi(joint_prob::JointProbability, raw_prob::StringFloatMap)
  # Using formula p_pmi = max(0,log(P(context|targ)/P(context)))
  p_pmi = JointProbability()
  for (target::ASCIIString, context_probs::StringFloatMap) in joint_prob
    p_pmi[target] = StringFloatMap()
    for (context::ASCIIString, p_con_given_target::Float64) in context_probs
      if !haskey(raw_prob, context)
        continue
      end

      p_con = raw_prob[context]
      ppmi = p_con_given_target/p_con |> log
      
      p_pmi[target][context] = max(0., ppmi)
    end
  end

  p_pmi
end


function cosim(h1::StringFloatMap, h2::StringFloatMap)
    all_keys = union(keys(h1), keys(h2)) |> unique |> sort
    num_keys = length(all_keys)
    v1 = zeros(Float64, num_keys)
    v2 = zeros(Float64, num_keys)
    for (i, k) in enumerate(all_keys)
      v1[i] = get(h1, k, 0.)
      v2[i] = get(h2, k, 0.)
    end
    n1 = norm(v1)
    n2 = norm(v2)

    (n1 == 0. || n2 == 0.) ? 0.0 : dot(v1/n1, v2/n2)
end


function topItems(s::ASCIIString, p_pmi::JointProbability)
  cosims::Array{Tuple{Float64, ASCIIString}} = @>> keys(p_pmi) begin
    filter(k -> k != s)
    map(k -> (cosim(p_pmi[k], p_pmi[s]), k))
    sort(rev=true)
  end
  cosims[end - 15: end]
end


function simpleRun(p_pmi::JointProbability)
  for word in [ "eagle", "socks", "jacket", "asparagus", "brussels_sprouts",
                "spaghetti", "lion", "fun", "arrive", "straight" ]
    println("$(word)")
    top15 = topItems(word, p_pmi)
    for (cosim::Float64, cosim_word::ASCIIString) in top15
      println("\t $(cosim_word) $cosim")
    end
  end
end
